src/main/java/com/ryandw11/structure/bottomfill/DefaultBottomFill.java

```diff
@@ -1,77 +1,121 @@
 
-    private Location minLoc;
-    private Location maxLoc;
-    private Location location;
     private Structure structure;
-    private int currentX;
-    private int currentY;
-    private int currentZ;
-    private BukkitTask bukkitTask;
+    private Location spawnLocation;
+    private Queue<BlockVector2> groundPlane;
+    private int minY;
+    private Material fillMaterial;
 
     @Override
-    public void performFill(Structure structure, Location spawnLocation, Location minLoc, Location maxLoc) {
-        if (structure.getBottomSpaceFill().getFillMaterial(spawnLocation.getBlock().getBiome()).isEmpty())
-            return;
+    public void performFill(Structure structure, Location spawnLocation, Location minLoc, Location maxLoc) throws IOException {
+        var fillMaterial = structure.getBottomSpaceFill().getFillMaterial(spawnLocation.getBlock().getBiome());
+        if (fillMaterial.isPresent()) {
+            this.fillMaterial = fillMaterial.get();
+        } else return;
 
         this.structure = structure;
-        this.location = spawnLocation;
-        this.minLoc = new Location(minLoc.getWorld(), minLoc.getBlockX(), minLoc.getBlockY() - 1, minLoc.getBlockZ());
-        this.maxLoc = new Location(minLoc.getWorld(), maxLoc.getBlockX(), minLoc.getBlockY() - 1, maxLoc.getBlockZ());
-        currentX = minLoc.getBlockX();
-        currentY = minLoc.getBlockY() - 1;
-        currentZ = minLoc.getBlockZ();
-        bukkitTask = this.runTaskTimer(CustomStructures.getInstance(), 0, 1);
+        this.spawnLocation = spawnLocation;
+        this.minY = minLoc.getBlockY();
+
+        // The 2D plane which the blocks will be ground placed on
+        groundPlane = new LinkedList<>();
+
+        // The world to manipulate
+        World world = spawnLocation.getWorld();
+        Preconditions.checkNotNull(world, "world");
+
+        // To get the ground plane, we need to read the schematic
+        File file = new File(CustomStructures.getInstance().getDataFolder() + "/schematics/" + structure.getSchematic());
+        ClipboardFormat format = ClipboardFormats.findByFile(file);
+        if (format == null) {
+            CustomStructures.getInstance().getLogger().warning("Invalid schematic format for schematic " + structure.getSchematic() + "!");
+            CustomStructures.getInstance().getLogger().warning("Please create a valid schematic using the in-game commands!");
+            return;
+        }
+        try (ClipboardReader reader = format.getReader(new FileInputStream(file))) {
+            Clipboard clipboard = reader.read();
+
+            // Note: The minimum point of a region is the lowest abs point in the original world
+
+            var minX = minLoc.getBlockX();
+            var minZ = minLoc.getBlockZ();
+            var clipboardMinY = clipboard.getMinimumPoint().getBlockY();
+            for (int x = clipboard.getMinimumPoint().getBlockX(); x <= clipboard.getMaximumPoint().getBlockX(); x++) {
```

Comment:
This code does not factor in a random rotation.

Reply:
Does it mean I need to rotate the schematic first according to what is configured in the strcuture config, before finding the ground points?

Reply:
Yes, you need to rotate the schematic first.